12.0.x: Rework `consumeAvailable()` mechanism

lorban opened the pull request:
Started with a fix for a comment detailing the `consumeAvailable()` behavior that is partially incorrect, ended up reworking the `consumeAvailable()` mechanism.

See https://github.com/jetty/jetty.project/pull/13934#discussion_r2500556624

See https://github.com/jetty/jetty.project/pull/14026 for the 12.1.x version of this PR.

lorban committed:
fix comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw reviewed (changes requested):
I'm a bit more concerned about this now, as I didn't realise that `HttpStreamOverHttp1#consumeAvailable` made the connection non-persistent.

If this is the case, it is a little bit redundant as we have the method org.eclipse.jetty.server.ResponseUtils#ensureConsumeAvailableOrNotPersistent, which does a much better job of being non persistent.

I think we need to look a little closer before moving on, and maybe document a b it more

lorban committed:
tentative: bypass ensureNotPersistent call

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
> I'm a bit more concerned about this now, as I didn't realise that `HttpStreamOverHttp1#consumeAvailable` made the connection non-persistent.
> 
> If this is the case, it is a little bit redundant as we have the method org.eclipse.jetty.server.ResponseUtils#ensureConsumeAvailableOrNotPersistent, which does a much better job of being non persistent.
> 
> I think we need to look a little closer before moving on, and maybe document a b it more

`ResponseUtils#ensureConsumeAvailableOrNotPersistent` works at the `Request` level, so eventually that work is delegated to `HttpStream#consumeAvailable`.

 Out of curiosity, I tried disabling the `ensureNotPersistent` call from `ensureConsumeAvailableOrNotPersistent` and that didn't break a single test, so I'm tempted to believe `ResponseUtils#ensureConsumeAvailableOrNotPersistent` could be replaced with `Request#consumeAvailable`, and `ResponseUtils` removed.

lorban committed:
tentative: remove ResponseUtils

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
tentative: remove ResponseUtils, add javadoc, add todos

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
I pushed the removal of `ResponseUtils` and added the missing javadoc.

How does that look?

lorban committed:
fix consumeAvailable() when request is wrapped + add test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw I've reworked the whole `consumeAvailable()` mechanism to get rid of `ResponseUtils`, and fixed `ServletCoreRequest.consumeAvailable()` to make sure it never blocks even when the servlet request is wrapped.

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResponseUtils.java:
```diff
@@ -1,86 +0,0 @@
-                response.getHeaders().computeField(HttpHeader.CONNECTION, (h, fields) ->
-                {
-                    if (fields == null || fields.isEmpty())
-                        return null;
-                    String v = fields.stream()
-                        .flatMap(field -> Stream.of(field.getValues()).filter(s -> !HttpHeaderValue.KEEP_ALIVE.is(s)))
-                        .collect(Collectors.joining(", "));
-                    if (StringUtil.isEmpty(v))
-                        return null;
-
-                    return new HttpField(HttpHeader.CONNECTION, v);
-                });
-
-            case HTTP_1_1 ->
-                // Add close value to Connection headers
-                response.getHeaders().computeField(HttpHeader.CONNECTION, (h, fields) ->
-                {
-                    if (fields == null || fields.isEmpty())
-                        return HttpFields.CONNECTION_CLOSE;
-
-                    if (fields.stream().anyMatch(f -> f.contains(HttpHeaderValue.CLOSE.asString())))
-                    {
-                        if (fields.size() == 1)
-                        {
-                            HttpField f = fields.get(0);
-                            if (HttpFields.CONNECTION_CLOSE.equals(f))
-                                return f;
-                        }
-
-                        return new HttpField(HttpHeader.CONNECTION, fields.stream()
-                            .flatMap(field -> Stream.of(field.getValues()).filter(s -> !HttpHeaderValue.KEEP_ALIVE.is(s)))
-                            .collect(Collectors.joining(", ")));
-                    }
-
-                    return new HttpField(HttpHeader.CONNECTION,
-                        Stream.concat(fields.stream()
-                                    .flatMap(field -> Stream.of(field.getValues()).filter(s -> !HttpHeaderValue.KEEP_ALIVE.is(s))),
-                                Stream.of(HttpHeaderValue.CLOSE.asString()))
-                            .collect(Collectors.joining(", ")));
-                });
```
I think this is good... but I need to ponder for a bit.    I believe this logic is actually all contained within the HttpGenerator,  so that removing this method gets rid of duplication... but I need to convince myself that it is all there.  I'll take another day to walk through all the variations

gregw replied:
This logic is indeed done entirely by HttpGenerator.

gregw reviewed (commented):
Thinking on this one for a day...

gregw reviewed (approved):
This is good.   The HttpGenerator does all the headers required.
